libvast/src/table_slice.cpp

```diff
@@ -823,4 +815,255 @@ auto resolve_operand(const table_slice& slice, const operand& op)
+
+auto flatten_record(
+  std::string_view separator, std::string_view name_prefix,
+  const std::vector<std::shared_ptr<arrow::Array>>& list_offsets,
+  struct record_type::field field, const std::shared_ptr<arrow::Array>& array)
+  -> std::vector<
+    std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>>;
+
+auto flatten_list(std::string_view separator, std::string_view name_prefix,
+                  std::vector<std::shared_ptr<arrow::Array>> list_offsets,
+                  struct record_type::field field,
+                  const std::shared_ptr<arrow::Array>& array)
+  -> std::vector<
+    std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+  using result_type = std::vector<
+    std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>>;
+  const auto& lt = caf::get<list_type>(field.type);
+  auto list_array
+    = std::static_pointer_cast<type_to_arrow_array_t<list_type>>(array);
+  list_offsets.push_back(list_array->offsets());
+  auto f = detail::overload{
+    [&]<concrete_type Type>(const Type&) -> result_type {
+      auto result = result_type{};
+      if (list_offsets.empty()) {
+        result.push_back({
+          {
+            fmt::format("{}{}", name_prefix, field.name),
+            field.type,
+          },
+          array,
+        });
+      } else {
+        auto combined_offsets = combine_offsets(list_offsets);
+        result.push_back({
+          {
+            fmt::format("{}{}", name_prefix, field.name),
+            field.type,
+          },
+          arrow::ListArray::FromArrays(*combined_offsets, *list_array->values())
+            .ValueOrDie(),
+        });
+      }
+      return result;
+    },
+    [&](const list_type& lt) -> result_type {
+      return flatten_list(separator, name_prefix, std::move(list_offsets),
+                          {field.name, lt}, list_array->values());
+    },
+    [&](const record_type& rt) -> result_type {
+      return flatten_record(separator, name_prefix, list_offsets,
+                            {field.name, rt}, list_array->values());
+    },
+  };
+  return caf::visit(f, lt.value_type());
+}
+
+auto flatten_record(
+  std::string_view separator, std::string_view name_prefix,
+  const std::vector<std::shared_ptr<arrow::Array>>& list_offsets,
+  struct record_type::field field, const std::shared_ptr<arrow::Array>& array)
+  -> std::vector<
+    std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+  using result_type = std::vector<
+    std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>>;
+  const auto& rt = caf::get<record_type>(field.type);
+  auto struct_array
+    = std::static_pointer_cast<type_to_arrow_array_t<record_type>>(array);
+  const auto next_name_prefix
+    = fmt::format("{}{}{}", name_prefix, field.name, separator);
+  auto transformations = std::vector<indexed_transformation>{};
+  for (size_t i = 0; i < rt.num_fields(); ++i) {
+    transformations.emplace_back(
```

Comment:
This leads to a build error on my machine with Clang 15.0.7 - `no matching function for call to 'construct_at'`. `push_back` with an initializer list works.

Using GCC 12.2.1, this line works for me.

Also on lines 1020 & 1052.